src/NHibernate/Mapping/Any.cs

```diff
@@ -30,29 +31,47 @@ public virtual string IdentifierTypeName
 			set { identifierTypeName = value; }
 		}
 
-		private IType type;
-		public override IType Type
+		private Lazy<IType> GetLazyType()
 		{
-			get
-			{
-				if (type == null)
+			return new Lazy<IType>(
+				() =>
 				{
-					type =
-						new AnyType(
-							metaValues == null
-							? ("class".Equals(metaTypeName) ? new ClassMetaType(): TypeFactory.HeuristicType(metaTypeName))
-								: new MetaType(metaValues, TypeFactory.HeuristicType(metaTypeName)),
-							TypeFactory.HeuristicType(identifierTypeName));
-				}
-				return type;
-			}
+					var identifierType = TypeFactory.HeuristicType(identifierTypeName);
+					if (identifierType == null)
+						throw new MappingException($"Identifier type '{identifierTypeName}' is invalid");
+
+					var hasMetaValues = metaValues != null && metaValues.Count > 0;
+
+					if (!hasMetaValues && "class" == metaTypeName)
+						return new AnyType(new ClassMetaType(), identifierType);
+
+					var metaType = TypeFactory.HeuristicType(metaTypeName);
+					if (metaType == null)
+						throw new MappingException($"Meta type '{metaTypeName}' is invalid");
+
+					if (!hasMetaValues && metaType.ReturnedClass != typeof(string))
+						throw new MappingException(
+							$"Using a non-string meta type ('{metaTypeName}') without providing meta values is invalid");
+
+					return new AnyType(
+						hasMetaValues
+							? new MetaType(metaValues, TypeFactory.HeuristicType(metaTypeName))
+							: (IType) new ClassMetaType(metaType),
+						identifierType);
+				});
 		}
 
+		// Types may be used by many threads, we must use a thread safe lazy initialization.
+		// (Or we should build this class in another way, but this would likely be a breaking change.)
+		private Lazy<IType> _type;
```

Comment:
Unrelated fix by the way. If done separately, it would have to wait for the merge here (or cause conflicts).